### DISCUSSION TITLE
JavaScript work in added blocks via uib-element

### DISCUSSION THREAD

**Discussion by vovadpua** [TotallyInformation/node-red-contrib-uibuilder#220] [Q&A] [answered]
Good afternoon, I'm trying to figure out one question, since I'm new to programming, I don't know how to solve it. I am building a web page with blocks of html code using uib-element, I can form a page with three, four or more blocks. But when a block of code needs to perform a dynamic change, such as a slider that executes a JavaScript file, I can't get it to work dynamically. Please tell me where I can find out how to make JavaScript work in those blocks that were added via uib-element. Thanks in advance.

**Comment by vovadpua** [comment]
When the code is initially in the index.html file in which the links to the JavaScript file are written, then it works as expected, the only problem is to make it work in the newly added code through uib-element

**Comment by TotallyInformation** [comment]
Hi, at present, the zero-code features for user input relies on creating a form and using a button to send the form data back to Node-RED.

Individual and fully dynamic inputs for `uib-element` are on the roadmap.

So currently, create a form and use the send button to get the form data back into Node-RED.

![image](https://github.com/TotallyInformation/node-red-contrib-uibuilder/assets/1591850/de13c8d0-599f-4931-a025-e6f72d4213ea)

The zero-code example in the library has a form flow.

---

To get something more dynamic, you would need to use a generated low-code msg. For example, you could use `uib-element` to create an example form but instead of sending it to your uibuilder node, capture the output. Then you can look at the `msg._ui` property to see how the form is created. You could then remove the buttons and add an `onchange` attribute to each form element. 

As an example. Here is a msg.payload input to a `uib-element` node set to produce a form.

```
[
    {
        "id": "r1-text",
        "type": "text",
        "label": "Text Input:",
        "value": "Foo"
    },
    {
        "id": "r2-range",
        "type": "range",
        "label": "Range (0-100):",
        "value": "50",
        "min": 0,
        "max": 100
    }
]
```
The output from the `uib-element` node will look like this:
```
{
    "topic": "sform1",
    "mode": "update",
    "_ui": [
        {
            "method": "replace",
            "components": [
                {
                    "type": "div",
                    "id": "sform1",
                    "parent": "body",
                    "position": "last",
                    "attributes": {},
                    "components": [
                        {
                            "type": "form",
                            "components": [
                                {
                                    "type": "div",
                                    "components": [
                                        {
                                            "type": "label",
                                            "attributes": {
                                                "for": "r1-text"
                                            },
                                            "slot": "Text Input:"
                                        },
                                        {
                                            "type": "input",
                                            "id": "r1-text",
                                            "attributes": {
                                                "id": "r1-text",
                                                "type": "text",
                                                "label": "Text Input:",
                                                "value": "Foo",
                                                "onchange": "this.dataset.newValue = this.value",
                                                "onfocus": "this.dataset.oldValue = this.value",
                                                "name": "r1-text",
                                                "title": "Type: text"
                                            }
                                        }
                                    ],
                                    "attributes": {
                                        "title": "Type: text"
                                    }
                                },
                                {
                                    "type": "div",
                                    "components": [
                                        {
                                            "type": "label",
                                            "attributes": {
                                                "for": "r2-range"
                                            },
                                            "slot": "Range (0-100):"
                                        },
                                        {
                                            "type": "input",
                                            "id": "r2-range",
                                            "attributes": {
                                                "id": "r2-range",
                                                "type": "range",
                                                "label": "Range (0-100):",
                                                "value": "50",
                                                "min": 0,
                                                "max": 100,
                                                "onchange": "this.dataset.newValue = this.value",
                                                "onfocus": "this.dataset.oldValue = this.value",
                                                "name": "r2-range",
                                                "title": "Type: range"
                                            }
                                        },
                                        {
                                            "type": "div",
                                            "attributes": {},
                                            "components": [
                                                {
                                                    "type": "button",
                                                    "id": "sform1-btn-send",
                                                    "attributes": {
                                                        "type": "button",
                                                        "title": "Send the form data back to Node-RED"
                                                    },
                                                    "events": {
                                                        "click": "uibuilder.eventSend"
                                                    },
                                                    "slot": "Send"
                                                },
                                                {
                                                    "type": "button",
                                                    "id": "sform1-btn-reset",
                                                    "attributes": {
                                                        "type": "button",
                                                        "title": "Reset the form"
                                                    },
                                                    "events": {
                                                        "click": "event.srcElement.form.reset"
                                                    },
                                                    "slot": "Reset"
                                                }
                                            ]
                                        }
                                    ],
                                    "attributes": {
                                        "title": "Type: range"
                                    }
                                }
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}
```
Which probably looks a bit horrendous to begin with. But remember that HTML is hierarchical so work through each section looking for `components` arrays - these define the next level down in the hierarchy. This is a picture captured from the VScode editor showing a formatted and folder version of the above output and showing that the form itself is wrapped in a `div` and that it contains 2 direct children. 

![image](https://github.com/TotallyInformation/node-red-contrib-uibuilder/assets/1591850/4b94bd6d-3d62-4862-9b06-62a47a2790bb)

Each of those children represents 1 input. Each input contains another `div` wrapper which, in turn contains a label and an input. You are interested in the 2nd child. The actual input field is the 2nd component of that div. `"type": "input",` `"id": "r2-range",`. Expanding the attributes for that entry shows this:

![image](https://github.com/TotallyInformation/node-red-contrib-uibuilder/assets/1591850/179f427d-af84-42c9-946a-828d51e25629)

And to enable a field to dynamically send back updates, you would need to change the `onchange` attribute value to something like:

```
"onchange": "uibuilder.eventSend(event)",
```

Which will fire the function to get the form data back to Node-RED every time the user changes the value. The `onfocus` entry would be removed.

You can do the changes either using a function node with JavaScript or using a change node. Use the output of the `uib-element` node shown in the debug panel and expand that until you find the right element. You can then copy the path to that element and use that path directly in a change node.

![image](https://github.com/TotallyInformation/node-red-contrib-uibuilder/assets/1591850/213e2b48-d2f1-4086-b481-44a57a451f17)

![image](https://github.com/TotallyInformation/node-red-contrib-uibuilder/assets/1591850/9770a52a-846e-4bf7-9b1b-226d75a6acd1)

Here is an example flow snippet that illustrates all of the above.

```
[{"id":"c1a1873b92044750","type":"inject","z":"3badb0a6906eef7f","g":"a80a3c8e16edd295","name":"Add a Form","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"sform1","payload":"[{\"id\":\"r1-text\",\"type\":\"text\",\"label\":\"Text Input:\",\"value\":\"Foo\"},{\"id\":\"r2-range\",\"type\":\"range\",\"label\":\"Range (0-100):\",\"value\":\"50\",\"min\":0,\"max\":100}]","payloadType":"json","x":190,"y":640,"wires":[["889250edc1409ac3"]],"info":"Creates an example form showing off different features.\r\n\r\nNo buttons are included so that the `uib-element` node\r\nautomatically creates the two \"Send\" and \"Reset\" default\r\nbuttons for you.\r\n\r\nEmail & Data inputs are marked as required but left with\r\nno default value to show the invalid field/form highlights."},{"id":"889250edc1409ac3","type":"uib-element","z":"3badb0a6906eef7f","g":"a80a3c8e16edd295","name":"","topic":"","elementtype":"sform","parent":"body","parentSource":"","parentSourceType":"str","elementid":"sform1","elementId":"","elementIdSourceType":"str","heading":"","headingSourceType":"str","headingLevel":"h2","position":"last","positionSourceType":"str","confData":{},"x":450,"y":640,"wires":[["d979be11fd2983d0","68a71f6a04286830"]]},{"id":"d979be11fd2983d0","type":"debug","z":"3badb0a6906eef7f","g":"a80a3c8e16edd295","name":"debug 29","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","targetType":"full","statusVal":"","statusType":"auto","x":575,"y":660,"wires":[],"l":false},{"id":"68a71f6a04286830","type":"change","z":"3badb0a6906eef7f","g":"a80a3c8e16edd295","name":"","rules":[{"t":"set","p":"_ui[0].components[0].components[0].components[1].components[1].attributes.onchange","pt":"msg","to":"uibuilder.eventSend(event)","tot":"str"},{"t":"delete","p":"_ui[0].components[0].components[0].components[1].components[2]","pt":"msg"}],"action":"","property":"","from":"","to":"","reg":false,"x":720,"y":640,"wires":[["6c9106c11abceb5d"]]}]
```

Once you've got your head around the basics of manipulating the low-code config data, you will hopefully begin to see how you can create your own custom outputs. For example, if you want to have immediate outputs from all user inputs, you don't actually need to wrap everything in a form.

**Comment by TotallyInformation** [comment]
There are, of course, other ways to achieve this. You could write some front-end JavaScript in `index.js` that would add event handling to any input field on-screen. Which is what I think you were leaning towards. However, if you aren't familiar with JavaScript, you may prefer to keep things in Node-RED?

If you want an example of doing the processing in the browser though, let me know and I'll knock something up.

**Reply by vovadpua** [reply]
Thank you for your quick response. I had a little different. I have code [codepen](https://codepen.io/Vladimir-Vladimir-the-sans/pen/BaGyVwB) html css, javascript, if I place it in index.html, index.css, index.js files, then everything works correctly for me. But I want to write the html code in the template node, and send it via uib-element type html, so that it works the same way as in the index.html file.
As I understand it, when a web page is loading, an empty index.html is loaded with css styles and js code attached, but when I then inject the html code, it does not interact with the js code. 
I Tried different ways to make JS work after html code injection
[](url)
HTML:
   `<div>...</div>`
  ` <script>`
 `   //my script`
   `</script>`


or this code:
JS:
function myFunc() {
   //my script
};
 HTML:
`<div onload="myFunc()">...</div>`

but nothing worked.
Excuse my bad english, I am trying to explain how I present it.
Thank you.

**Reply by TotallyInformation** [reply]
OK, interesting. Let me have a go, I'd not actually thought of that approach. Though it should work. There is another method to do that from Node-RED as well, let me experiment.

**Reply by TotallyInformation** [reply, marked as answer]
Oops, I remembered why you can't do this the way you were hoping. It is because, while uibuilder does happily add the script tag and its content, that script is not executed because it was loaded after the page finished loading.

So there is no way to be able to do this currently using the zero-code `uib-element` node. However, this is Node-RED so there are always multiple ways to achieve things and you can still do it with minimal effort. 😊

The trick being to use a change node instead of a uib-element node. You will need to know how to load the script dynamically and this can be done using uibuilder's low-code features with the following JSON - just use the change node to set the contents of `msg._ui` using JSONata. This is the code:

```json
{
    "method": "load",
    "txtScripts": [
        payload
    ]
}
```

Note the `payload`. JSONata will replace that with the output from your template node.

The load method is documented here: https://totallyinformation.github.io/node-red-contrib-uibuilder/#/client-docs/config-driven-ui?id=method-load.

Just note the limitations of this approach. It will take some time for your browser to convert the text to code and execute it. This means that if, for example, you were expecting to be able to, in the same flow, follow this with something that relied on the code, you should send this and then have a small delay before sending the reliant data.

Here is a very simple flow fragment:

```
[{"id":"1f6feb8810af78a4","type":"inject","z":"30fdd9a9702231b0","name":"","props":[{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"test dynamic script load","x":135,"y":1080,"wires":[["698fc18e262ff6f3"]],"l":false},{"id":"698fc18e262ff6f3","type":"template","z":"30fdd9a9702231b0","name":"","field":"payload","fieldType":"msg","format":"handlebars","syntax":"mustache","template":"function hello() {\n    console.log('hello')\n}\n\n// This will be executed immediately on submission\nhello()\n","output":"str","x":240,"y":1080,"wires":[["5cdd3cf321df8625"]]},{"id":"5cdd3cf321df8625","type":"change","z":"30fdd9a9702231b0","name":"","rules":[{"t":"set","p":"_ui","pt":"msg","to":"{\t    \"method\": \"load\",\t    \"txtScripts\": [\t        payload\t    ]\t}\t","tot":"jsonata"}],"action":"","property":"","from":"","to":"","reg":false,"x":450,"y":1080,"wires":[["33ee5ceec2370977","6fa3635abe13dfed"]]},{"id":"6fa3635abe13dfed","type":"debug","z":"30fdd9a9702231b0","name":"debug 120","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"_ui","targetType":"msg","statusVal":"","statusType":"auto","x":630,"y":1100,"wires":[]}]
```

**Reply by TotallyInformation** [reply]
Just checking your example code. 

Personally, I would leave that code in the index.css and index.js files, it is a cleaner solution over all. But it works either way.

If you need to dynamically load the CSS, add another template node for just the CSS but change its output to something other than `msg.payload` then you can include it in the same flow segment simply by extending the JSONata in the change node:

```json
{
    "method": "load",
    "txtScripts": [
        payload
    ],
    "txtStyle": [
        payload
    ]
}
```

Again though, remember that you are late-loading the styles and so you may get a flash of unstyled content.

Unless you have a real need to make the input code dynamic, it is best left in the front-end code.

**Reply by vovadpua** [reply]
Thank you very much for your answer, I will study your recommendations.
